### PR TITLE
For local file copy create a manifest with a list of files.

### PR DESCRIPTION
**Pull request by ekini**
That allows to safely delete only the files we copied on the previous
copy, and not worry about stale files still living in the working directory.

Also removes some unused functions.

Initially I wanted to compare source and destination directories, but we use [CopyFolderContentsWithFilter](https://github.com/gruntwork-io/terragrunt/blob/master/util/file.go#L150) at least two times, so this will not work.

The only way I see is to create a list of files we touch, and delete them just before copying new files.

Used [gob](https://golang.org/pkg/encoding/gob) as the simplest serialiser. Added some tests.

### CODE REVIEW

File: cli/download_source.go [hunk 1]
```diff
@@ -57,7 +56,7 @@ func downloadTerraformSource(source string, terragruntOptions *options.Terragrun
 	}
 
 	terragruntOptions.Logger.Printf("Copying files from %s into %s", terragruntOptions.WorkingDir, terraformSource.WorkingDir)
-	if err := util.CopyFolderContents(terragruntOptions.WorkingDir, terraformSource.WorkingDir); err != nil {
+	if err := util.CopyFolderContents(terragruntOptions.WorkingDir, terraformSource.WorkingDir, ".terragrunt-module-manifest"); err != nil {
```
Comment (brikis98) [inline cli/download_source.go]: `".terragrunt-module-manifest"` should be stored in a `const`

File: util/file.go [hunk 2]
```diff
@@ -139,18 +131,26 @@ func ReadFileAsString(path string) (string, error) {
 
 // Copy the files and folders within the source folder into the destination folder. Note that hidden files and folders
 // (those starting with a dot) will be skipped.
-func CopyFolderContents(source string, destination string) error {
-	return CopyFolderContentsWithFilter(source, destination, func(path string) bool {
+func CopyFolderContents(source, destination, manifestFile string) error {
```
Comment (brikis98) [inline util/file.go]: Comment on function should be updated

File: util/file.go [hunk 3]
```diff
@@ -139,18 +131,26 @@ func ReadFileAsString(path string) (string, error) {
...
+	return CopyFolderContentsWithFilter(source, destination, manifestFile, func(path string) bool {
 		return !PathContainsHiddenFileOrFolder(path)
 	})
 }
 
 // Copy the files and folders within the source folder into the destination folder. Pass each file and folder through
 // the given filter function and only copy it if the filter returns true.
-func CopyFolderContentsWithFilter(source string, destination string, filter func(path string) bool) error {
+func CopyFolderContentsWithFilter(source, destination, manifestFile string, filter func(path string) bool) error {
```
Comment (brikis98) [inline util/file.go]: Comment on function should be updated

File: util/file.go [hunk 4]
```diff
@@ -139,18 +131,26 @@ func ReadFileAsString(path string) (string, error) {
...
 	// Why use filepath.Glob here? The original implementation used ioutil.ReadDir, but that method calls lstat on all
 	// the files/folders in the directory, including files/folders you may want to explicitly skip. The next attempt
 	// was to use filepath.Walk, but that doesn't work because it ignores symlinks. So, now we turn to filepath.Glob.
```
Comment (brikis98) [inline util/file.go]: Comment should be moved above actual `filepath.Glob` call

File: util/file.go [hunk 5]
```diff
@@ -189,10 +189,13 @@ func CopyFolderContentsWithFilter(source string, destination string, filter func
 			if err := CopyFile(file, dest); err != nil {
 				return err
 			}
+			if err := manifest.AddFile(dest); err != nil {
```
Comment (brikis98) [inline util/file.go]: The manifest file doesn't track folders? Just files?

File: util/file.go [hunk 6]
```diff
@@ -253,20 +256,68 @@ func JoinTerraformModulePath(modulesFolder string, path string) string {
...
+	Path       string
+	encoder    *gob.Encoder
+	fileHandle *os.File
+}
 
-	for _, pathPart := range pathParts {
-		if pathPart == folderName {
-			return true
+// Clean will remove all files specified in the manifest
+func (f *fileManifest) Clean() error {
```
Comment (brikis98) [inline util/file.go]: Please don't use single-letter variable names. They are hard to follow more than 1-2 lines away from the declaration.

File: util/file.go [hunk 7]
```diff
@@ -253,20 +256,68 @@ func JoinTerraformModulePath(modulesFolder string, path string) string {
...
+
+	// if manifest file doesn't exist, just exit
+	if !FileExists(f.Path) {
+		return nil
+	}
+	file, err := os.Open(f.Path)
+	if err != nil {
+		return err
+	}
+	decoder := gob.NewDecoder(file)
```
Comment (brikis98) [inline util/file.go]: Sorry, I'm probably missing something obvious, but what is the role of the `decoder` here?

File: util/file.go [hunk 8]
```diff
@@ -189,10 +189,13 @@ func CopyFolderContentsWithFilter(source string, destination string, filter func
...
 				return err
 			}
+			if err := manifest.AddFile(dest); err != nil {
+				return err
+			}
 		}
 	}
 
-	return nil
+	return manifest.Close()
```
Comment (brikis98) [inline util/file.go]: Should this be in a `defer` to ensure the file is closed even if we hit an error part way through?

File: util/file_test.go [hunk 9]
```diff
@@ -120,31 +121,27 @@ func TestJoinTerraformModulePath(t *testing.T) {
...
-	for _, testCase := range testCases {
-		// The following is necessary to make sure testCase's values don't
-		// get updated due to concurrency within the scope of t.Run(..) below
-		testCase := testCase
-		t.Run(testCase.description, func(t *testing.T) {
-			actual := PathContains(testCase.path, testCase.folderName)
-			assert.Equal(t, testCase.expected, actual)
-		})
+	manifest := newFileManifest(".terragrunt-test-manifest")
+	assert.Nil(t, manifest.Create())
```
Comment (brikis98) [inline util/file_test.go]: `require.Nil`

File: util/file_test.go [hunk 10]
```diff
@@ -120,31 +121,27 @@ func TestJoinTerraformModulePath(t *testing.T) {
...
+	assert.FileExists(t, manifest.Path)
+	for _, file := range testfiles {
+		assert.Nil(t, manifest.AddFile(file))
+	}
+
+	assert.Nil(t, manifest.Clean())
```
Comment (brikis98) [inline util/file_test.go]: `require.Nil`

File: util/file_test.go [hunk 11]
```diff
@@ -120,31 +121,27 @@ func TestJoinTerraformModulePath(t *testing.T) {
...
-		expected    bool
-	}{
-		{"both empty", "", "", true},
-		{"path empty, folder name not empty", "", "foo", false},
-		{"path non empty, folder name empty", "/foo", "", true},
-		{"path contains one part that matches folder name", "foo", "foo", true},
-		{"path contains one part that doesn't match folder name", "foo", "bar", false},
-		{"path contains multiple parts, one of which matches folder name", "/foo/bar/baz", "bar", true},
-		{"path contains multiple parts, none of which match folder name", "/foo/bar/baz", "will-not-match", false},
+	var testfiles = []string{"file1", "file2"}
```
Comment (brikis98) [inline util/file_test.go]: Thanks for adding tests! 👍 

However, I'd strongly recommend having tests for all the corner cases, as those are what'll kill us here:

1. Empty folder
1. Nested files/folders
1. Nested folders, some of which are empty
1. Etc

File: cli/file_copy_getter.go [hunk 12]
```diff
@@ -30,7 +31,7 @@ func (g *FileCopyGetter) Get(dst string, u *url.URL) error {
 		return fmt.Errorf("source path must be a directory")
 	}
 
-	return util.CopyFolderContents(path, dst)
+	return util.CopyFolderContents(path, dst, ".terragrunt-source-manifest")
```
Comment (brikis98) [inline cli/file_copy_getter.go]: Shouldn't this be an absolute path, not a relative one? It seems important that the manifest file is stored in the proper folder!

File: util/file.go [hunk 13]
```diff
@@ -253,20 +256,68 @@ func JoinTerraformModulePath(modulesFolder string, path string) string {
...
 
-// Returns true if the given path contains the given folder name.
-//
-// Examples:
-//
-// PathContains("/foo/bar", "foo") => returns true
-// PathContains("/foo/bar", "baz") => returns false
-func PathContains(path string, folderName string) bool {
-	pathParts := strings.Split(path, string(filepath.Separator))
+type fileManifest struct {
```
Comment (brikis98) [inline util/file.go]: Please add comments explaining what this is, why it exists, and how it's meant to be used.

File: util/file.go [hunk 14]
```diff
@@ -253,20 +256,68 @@ func JoinTerraformModulePath(modulesFolder string, path string) string {
...
+	// remove the manifest itself
+	if err := os.RemoveAll(f.Path); err != nil {
+		return errors.WithStackTrace(err)
+	}
 
-	return false
+	return nil
+}
+
+func (f *fileManifest) Create() error {
```
Comment (brikis98) [inline util/file.go]: Please add comments to these methods

File: cli/download_source.go (hunk 1 above)
Comment (ekini) [reply cli/download_source.go]: Yes, I was thinking about this too. Sounds reasonable, will fix that

File: util/file.go (hunk 5 above)
Comment (ekini) [reply util/file.go]: Yes, because on directories, it calls `CopyFolderContentsWithFilter` recursively, and subfolders will have their own manifests.

File: util/file.go (hunk 7 above)
Comment (ekini) [reply util/file.go]: The whole point of `gob` is that it provides ability to decode filenames one by one, instead of load the file in memory at once.

File: cli/file_copy_getter.go (hunk 12 above)
Comment (ekini) [reply cli/file_copy_getter.go]: Oh, it's not a path, but a name. It's stored in the `dst` folder.

**Comment by ekini** [comment]
I've had another look and it seems the code didn't handle subfolders correctly. I'll fix that and add more tests with corner cases.

File: util/file.go (hunk 5 above)
Comment (ekini) [reply util/file.go]: Now it tracks directories as well.

File: util/file.go (hunk 13 above)
Comment (ekini) [reply util/file.go]: done :)

**Comment by ekini** [comment]
@brikis98 added more tests, fixed subfolders handling, looks much better to me now :)

**Comment by mattford63** [comment]
Hi, this doesn't seem to fix the issue I'm experiencing from the original issue.  I have to still clear the cache and the error produced is as per the originally reported bug.

https://github.com/gruntwork-io/terragrunt/issues/764#issuecomment-511704040

Sorry I can't help more at the moment.

**Comment by ekini** [comment]
Ok, found another bug when the destination directory is not created, but it tries to create a manifest file there. Fixed

**Comment by ekini** [comment]
I've created a homebrew formula that installs terragrunt with this patch, along with `terraform` version that doesn't complain about unset variables. Just if anyone wants to try it.
```bash
$brew install springload/tools/terragrunt
```

File: cli/file_copy_getter.go [hunk 15]
```diff
@@ -30,7 +34,7 @@ func (g *FileCopyGetter) Get(dst string, u *url.URL) error {
 		return fmt.Errorf("source path must be a directory")
 	}
 
-	return util.CopyFolderContents(path, dst)
+	return util.CopyFolderContents(path, dst, SOURCE_MANIFEST_NAME)
```
Comment (brikis98) [inline cli/file_copy_getter.go]: Should this be using a different constant than `MODULE_MANIFEST_NAME`?

File: util/file.go [hunk 16]
```diff
@@ -253,20 +264,110 @@ func JoinTerraformModulePath(modulesFolder string, path string) string {
...
+// This allows to clean those files on subsequent runs.
+// The problem is as follows: terragrunt copies the terraform source code first to "working directory" using go-getter,
+// and then copies all files from the directory it runs from to the above dir.
+// It works fine on the first run, but if we delete a file from the current terragrunt directory, we want it
+// to be cleaned in the "working directory" as well. Since we don't really know what can get copied by go-getter,
+// we have to track all the files we tough in a manifest. This way we know exactly which files we need to clean on
+// subsequent runs.
+type fileManifest struct {
+	Path         string
+	ManifestFile string
```
Comment (brikis98) [inline util/file.go]: What is `Path` vs `ManifestFile`? Is the former the folder path and the latter the path to the manifest file within that folder?

File: util/file.go [hunk 17]
```diff
@@ -253,20 +264,110 @@ func JoinTerraformModulePath(modulesFolder string, path string) string {
...
+func (manifest *fileManifest) clean(path string) error {
+	manifestPath := filepath.Join(path, manifest.ManifestFile)
+
+	// if manifest file doesn't exist, just exit
+	if !FileExists(manifestPath) {
+		return nil
+	}
+	file, err := os.Open(manifestPath)
+	if err != nil {
+		// the manifest doesn't exist, so there's nothing we can safely clean
```
Comment (brikis98) [inline util/file.go]: Didn't you check for manifest existing just one line above? Seems redundant to have the same check twice.

File: util/file.go [hunk 18]
```diff
@@ -253,20 +264,110 @@ func JoinTerraformModulePath(modulesFolder string, path string) string {
...
+			} else {
+				return err
+			}
+		}
+		if manifestEntry.IsDir {
+			if err := manifest.clean(manifestEntry.Path); err != nil {
+				return errors.WithStackTrace(err)
+			}
+		} else {
+			if err := os.RemoveAll(manifestEntry.Path); err != nil {
```
Comment (brikis98) [inline util/file.go]: If `manifestEntry` is a file path, shouldn't this use `os.Remove` instead of `os.RemoveAll`?

File: util/file.go [hunk 19]
```diff
@@ -253,20 +264,110 @@ func JoinTerraformModulePath(modulesFolder string, path string) string {
...
+				return errors.WithStackTrace(err)
+			}
+		}
+	}
+	if err := os.RemoveAll(manifestPath); err != nil {
```
Comment (brikis98) [inline util/file.go]: Same here: `os.Remove` for files?

File: util/file.go [hunk 20]
```diff
@@ -253,20 +264,110 @@ func JoinTerraformModulePath(modulesFolder string, path string) string {
...
+		return errors.WithStackTrace(err)
+	}
+	// remove the manifest itself
+	// it will run after the close defer
+	defer os.RemoveAll(manifestPath)
```
Comment (brikis98) [inline util/file.go]: Wait, the line above removes `manifestPath`... Why do it again?

File: util/file.go [hunk 21]
```diff
@@ -253,20 +264,110 @@ func JoinTerraformModulePath(modulesFolder string, path string) string {
...
+}
+
+// Create will create the manifest file
+func (manifest *fileManifest) Create() error {
+	var err error
+	manifest.fileHandle, err = os.OpenFile(filepath.Join(manifest.Path, manifest.ManifestFile), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
+	if err != nil {
+		return err
+	}
+	manifest.encoder = gob.NewEncoder(manifest.fileHandle)
```
Comment (brikis98) [inline util/file.go]: Very minor nit, but it seems slightly cleaner to do:

```go
fileHandle, err := os.OpenFile(...)
if err != nil {
  return err
}
manifest.fileHandle = fileHandle
manifest.encoder = gob.NewEncoder(...)
```

File: cli/file_copy_getter.go (hunk 15 above)
Comment (ekini) [reply cli/file_copy_getter.go]: Yes, because if a local source is used, then two separate manifest files need to be created: one for the local source, and the other for the terragrunt module directory.

File: util/file.go (hunk 18 above)
Comment (ekini) [reply util/file.go]: Yes, fixed that

File: util/file.go (hunk 17 above)
Comment (ekini) [reply util/file.go]: I must have been drunk. Fixed

File: util/file.go (hunk 16 above)
Comment (ekini) [reply util/file.go]: Renamed `Path` to `ManifestFolder` and added comments to each var. It used to be just `Path` before handling the subdirectories properly.

File: util/file.go (hunk 20 above)
Comment (ekini) [reply util/file.go]: removed the extra remove 👍

File: util/file.go (hunk 21 above)
Comment (ekini) [reply util/file.go]: Done :) Not sure why it's not showing as "Outdated" on github.

File: cli/file_copy_getter.go (hunk 15 above)
Comment (brikis98) [reply cli/file_copy_getter.go]: Ohhh, I gotcha. I'll update one comment to make that clearer.

File: cli/download_source.go [hunk 22]
```diff
@@ -16,9 +16,11 @@ import (
 	"github.com/gruntwork-io/terragrunt/util"
 	"github.com/hashicorp/go-getter"
 	urlhelper "github.com/hashicorp/go-getter/helper/url"
-	"github.com/mattn/go-zglob"
 )
 
+// manifest for files coped from terragrunt module folder
```
Comment (brikis98) [inline cli/download_source.go]:
```suggestion
// manifest for files coped from terragrunt module folder (i.e., the folder that contains the current terragrunt.hcl)
```

File: cli/file_copy_getter.go [hunk 23]
```diff
@@ -2,12 +2,16 @@ package cli
...
-	"github.com/gruntwork-io/terragrunt/util"
-	"github.com/hashicorp/go-getter"
 	"net/url"
 	"os"
+
+	"github.com/gruntwork-io/terragrunt/util"
+	"github.com/hashicorp/go-getter"
 )
 
+// manifest for files copied from terraform source
```
Comment (brikis98) [inline cli/file_copy_getter.go]:
```suggestion
// manifest for files copied from the URL specified in the terraform { source = "<URL>" } config
```

**Comment by brikis98** [comment]
Unit tests failed:

```
--- FAIL: TestDownloadTerraformSourceFromLocalFolderWithManifest (0.00s)
    --- PASS: TestDownloadTerraformSourceFromLocalFolderWithManifest/test-stale-file-exists (0.00s)
    --- PASS: TestDownloadTerraformSourceFromLocalFolderWithManifest/test-stale-file-doesnt-exist-after-source-update (0.00s)
    --- PASS: TestDownloadTerraformSourceFromLocalFolderWithManifest/test-tffile-exists-in-subfolder (0.00s)
    --- PASS: TestDownloadTerraformSourceFromLocalFolderWithManifest/test-tffile-doesnt-exist-in-subfolder (0.00s)
    --- FAIL: TestDownloadTerraformSourceFromLocalFolderWithManifest/test-empty-folder-gets-copied (0.00s)
        require.go:17: 
            	Error Trace:	download_source_test.go:213
            	Error:      	Condition failed!
            	Test:       	TestDownloadTerraformSourceFromLocalFolderWithManifest/test-empty-folder-gets-copied
```

**Comment by ekini** [comment]
@brikis98 I can't reproduce it. It works on my machine ™️ , plus I took a plain golang image:
```
docker run -ti --rm -v "$(pwd):/go/src/github.com/gruntwork/terragrunt" golang:1.12-alpine
/go # apk add git g++ gcc terraform > /dev/null && go get -u github.com/golang/dep/cmd/dep && cd src/github.com/gruntwork/terragrunt/
/go/src/github.com/gruntwork/terragrunt # dep ensure
/go/src/github.com/gruntwork/terragrunt # go test ./cli
ok      github.com/gruntwork/terragrunt/cli     29.499s
```
I'd be glad to debug the CircleCI build, but the docker image it uses seems to be private.

**Comment by brikis98** [comment]
Perhaps it's a race condition? Is `TestDownloadTerraformSourceFromLocalFolderWithManifest` copying lots of files/folders to a common destination and perhaps something is overwriting something else?

**Comment by ekini** [comment]
Doesn't seem to be a race condition. Not sure what's happening there. I haven't been able to reproduce it yet.

My wild guess is something wrong with temp folder in the Docker image.

**Comment by brikis98** [comment]
I'm probably missing something, but how is [this test case supposed](https://github.com/springload/terragrunt/blob/fix/file_copy/cli/download_source_test.go#L196-L200) to work? 

It looks like you're calling [copyFolder](https://github.com/springload/terragrunt/blob/fix/file_copy/cli/download_source_test.go#L212) on `"../test/fixture-manifest/version-5-empty-subfolder"`, which [solely contains a main.tf file](https://github.com/springload/terragrunt/tree/fix/file_copy/test/fixture-manifest/version-5-empty-subfolder). The test is then checking for a [sub2 folder](https://github.com/springload/terragrunt/blob/fix/file_copy/cli/download_source_test.go#L199). Where would that folder come from?

**Comment by ekini** [comment]
That's a very good point! It explains everything: when I tested it in a fresh Docker container I copied the codebase I had including the `sub2` folder, which obviously doesn't get committed to git. I've modified the test to create a temp folder with another folder inside, and use it as a source.

**Comment by brikis98** [comment]
OK, great, kicking off tests again

**Comment by brikis98** [comment]
Integration tests failed now:

```
Not on branch master, setting the short flag to skip long-running tests.
Running Go tests in /go/src/github.com/gruntwork-io/terragrunt/ with command: go test -v -timeout 45m -parallel 128 github.com/gruntwork-io/terragrunt/test github.com/gruntwork-io/terragrunt/test/helpers  -short
# github.com/gruntwork-io/terragrunt/test [github.com/gruntwork-io/terragrunt/test.test]
test/integration_test.go:35:2: assert redeclared as imported package name
	previous declaration at test/integration_test.go:21:2
test/integration_test.go:36:2: require redeclared as imported package name
	previous declaration at test/integration_test.go:22:2
test/integration_test.go:37:2: iterator redeclared as imported package name
	previous declaration at test/integration_test.go:23:2
FAIL	github.com/gruntwork-io/terragrunt/test [build failed]
?   	github.com/gruntwork-io/terragrunt/test/helpers	[no test files]
Exited with code 2
```

**Comment by ekini** [comment]
That was result of bad resolving of a conflict. I'm sure I've seen the CircleCI build process before, but can't see it now.

**Comment by brikis98** [comment]
Re-running tests.

**Comment by brikis98** [comment]
OK, all tests passed. Merging now. Thank you!

**Comment by brikis98** [comment]
https://github.com/gruntwork-io/terragrunt/releases/tag/v0.19.13
